ui/tests/integration/components/ttl-picker2-test.js

```diff
@@ -0,0 +1,117 @@
+    await render(hbs`
+      <TtlPicker2
+        @onChange={{onChange}}
+        @enableTTL={{true}}
+      />
+    `);
+
+    assert.dom('[data-test-ttl-value]').exists('TTL Picker time input exists');
+    assert.dom('[data-test-ttl-unit]').exists('TTL Picker unit select exists');
+  });
+
+  test('it does not show time and unit inputs when TTL disabled', async function(assert) {
+    let changeSpy = sinon.spy();
+    this.set('onChange', changeSpy);
+    await render(hbs`
+      <TtlPicker2
+        @onChange={{onChange}}
+        @enableTTL={{false}}
+      />
+    `);
+    assert.dom('[data-test-ttl-value]').doesNotExist('TTL Picker time input exists');
+    assert.dom('[data-test-ttl-unit]').doesNotExist('TTL Picker unit select exists');
+  });
+
+  test('it passes the appropriate data to onChange when toggled on', async function(assert) {
+    let changeSpy = sinon.spy();
+    this.set('onChange', changeSpy);
+    await render(hbs`
+      <TtlPicker2
+        @label="clicktest"
+        @unit="m"
+        @time="10"
+        @onChange={{onChange}}
+        @enableTTL={{false}}
+      />
+    `);
+    await click('[data-test-toggle-input="clicktest"]');
+    assert.ok(changeSpy.calledOnce, 'it calls the passed onChange');
+    assert.ok(
+      changeSpy.calledWith({
+        enabled: true,
+        seconds: 600,
+        timeString: '10m',
+      }),
+      'Passes the default values back to onChange'
+    );
+  });
+
+  test('it keeps seconds value when unit is changed', async function(assert) {
+    let changeSpy = sinon.spy();
+    this.set('onChange', changeSpy);
+    await render(hbs`
+      <TtlPicker2
+        @label="clicktest"
+        @unit="s"
+        @time="360"
+        @onChange={{onChange}}
+        @enableTTL={{false}}
+      />
+    `);
+    await click('[data-test-toggle-input="clicktest"]');
+    assert.ok(changeSpy.calledOnce, 'it calls the passed onChange');
+    assert.ok(
+      changeSpy.calledWith({
+        enabled: true,
+        seconds: 360,
+        timeString: '360s',
+      }),
+      'Passes the default values back to onChange'
+    );
+    await fillIn('[data-test-select="ttl-unit"]', 'm');
+    assert.ok(
+      changeSpy.calledWith({
+        enabled: true,
+        seconds: 360,
+        timeString: '6m',
+      }),
+      'Units and time update without changing seconds value'
+    );
+  });
+
+  test('it recalculates seconds when unit is changed and recalculateSeconds is on', async function(assert) {
+    let changeSpy = sinon.spy();
+    this.set('onChange', changeSpy);
+    await render(hbs`
+      <TtlPicker2
+        @label="clicktest"
+        @unit="s"
+        @time="120"
+        @onChange={{onChange}}
+        @enableTTL={{true}}
+        @recalculateSeconds={{true}}
+      />
+    `);
+    await fillIn('[data-test-select="ttl-unit"]', 'm');
+    assert.ok(
+      changeSpy.calledWith({
+        enabled: true,
+        seconds: 7200,
+        timeString: '120m',
```

Comment:
Great question, I was definitely getting dizzy while writing the description of how the tests should work, so I appreciate the feedback that the wording can be updated! 

From what I can tell, Ember waits until all the side effects of a function are complete before continuing with the next step of the test, so by default that means it's testing the case where unit is changed >TIMEOUT (5 seconds) after time is updated. Whether it behaves that way or in the pre-timeout way depends on the flag `recalculateSeconds` which is internally set in normal usage. Testing and concurrency aren't best friends yet, so to test the "within timeout" use case I manually set `recalculateSeconds` to true so that it would always behave the way it should when it's before timeout.

In terms of your specific use case, the "seconds" value isn't displayed anywhere on the UI, just passed back in the `ttl` object passed back in the onChange function. So the way I think about it is, if someone is changing the values one right after the other they probably have a good idea about what they want it set to (for example, "I want it to be 45 minutes" so they type 45 and then change the unit to minutes. The `seconds` value under the hood would change from 45 to 2700, while the `time` value (which is represented on the input) stays at 45 when the unit is changed. On the other hand, if I see `96 hours` on the input fields and, without touching the time input, change "hours" to "days" it should recalculate the `time` field to 4. 

I hope that helps a little bit! I can absolutely walk you through it in real time if you like 😄

Reply:
got it, thank you for the thorough explanation! that helps a ton!